src/sinks/new_relic/encoding.rs

```diff
@@ -0,0 +1,56 @@
+use crate::{
+    sinks::util::encoding::{
+        Encoder, EncodingConfigFixed
+    }
+};
+use serde::{
+    Deserialize, Serialize
+};
+use std::{
+    fmt::Debug,
+    io
+};
+use super::NewRelicApiModel;
+
+#[derive(Deserialize, Serialize, Debug, Eq, PartialEq, Clone, Derivative)]
+#[serde(rename_all = "snake_case")]
+#[derivative(Default)]
+pub enum Encoding {
+    #[derivative(Default)]
+    Default,
+}
+
+impl Encoder<Result<NewRelicApiModel, &'static str>> for EncodingConfigFixed<Encoding> {
+    fn encode_input(&self, input: Result<NewRelicApiModel, &'static str>, writer: &mut dyn io::Write) -> io::Result<usize> {
+        if let Ok(api_model) = input {
+            let json = match api_model {
+                NewRelicApiModel::Events(ev_api_model) => to_json(&ev_api_model),
+                NewRelicApiModel::Metrics(met_api_model) => to_json(&met_api_model),
+                NewRelicApiModel::Logs(log_api_model) => to_json(&log_api_model),
+            };
+            if let Some(json) = json {
+                let size = writer.write(&json)?;
+                io::Result::Ok(size)
+            }
+            else {
+                io::Result::Ok(0)
```

Comment:
The encoding should not silently fail here. An error message is still being printed (in `to_json`), but Vector still thinks this succeeded, and could cause empty requests to be sent. Instead, change the `to_json` to return `std::io::Result<Vec<u8>>` (which the `serde_json` Error can be converted into), and propagate that up in `encode_input` (which returns an `io::Result` already). That way, the error will be handled correctly.

Reply:
Same thing with the `Result<NewRelicApiModel, &'static str>` input. Any fallible logic should be handled earlier in the `sink.rs` stream, and only the `NewRelicApiModel` should be encoded.
You can see an example in the `splunk_hec/logs/sink.rs` `process_log` function, which does fallible processing in a `filter_map` to remove the failed events from the stream before it makes it to the request builder

Reply:
Resolved in https://github.com/vectordotdev/vector/pull/9764/commits/d312b84b9f3b92ad33fd261804ef485f5bfff0e6

Reply:
Regarding the propagation of errors. It's true that theoretically, an error could reach the request builder or even the encoder, but in practice, this is a very unlikely case, the error handling logic is only there because the `try_from` and the `to_json` method can fail, so these cases must be handled in-situ. But these situations are highly unexpected. In the first place, `NewRelicApiModel` creation only fails if all events in the Vec can't be converted, and this will only happen if the received events are really odd, for example receiving a LogEvent when a Metric is expected or the other way around. If only one of the events in the bulk can be converted, the API model is created correctly. In the second place, once the API model is created, I can't imagine a case in which Serde can't generate the JSON (and so, `to_json` can fail), because the format of an API model is always the same.

Having said that, I could rebuild the whole event handling chain to do something like what Splunk does, so if you really think that's something that can't be accepted the way it works now, I will rebuild it.